### PR TITLE
[Merged by Bors] - chore: Improved nth_rewrite and nth_rw docstrings

### PR DESCRIPTION
**Pull request by Shreyas4991**
[Zulip thread of discussion](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/docstrings.20of.20.60nth_rewrite.60.20and.20.60nth_rw.60/near/444853793)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/NthRewrite.lean
Comment (eric-wieser) [inline Mathlib/Tactic/NthRewrite.lean:46]:
```suggestion
`nth_rw n [eq₁, eq₂,...]` rewrites the `n`ᵗʰ occurrence of each equality `eq₁`, `eq₂`,... etc.
```
to match your other docstring.

File: Mathlib/Tactic/NthRewrite.lean
Comment (eric-wieser) [inline Mathlib/Tactic/NthRewrite.lean:22]:
```suggestion
to be rewritten. `nth_rewrite n [eq₁, eq₂, ..., eqₘ]` will rewrite the `n`ᵗʰ occurrence
of each of the `m` equalities `eqᵢ`.
```
(which is a bit like what the `simp` docstring uses)

File: Mathlib/Tactic/NthRewrite.lean
Comment (eric-wieser) [inline Mathlib/Tactic/NthRewrite.lean:26]:
```suggestion
Note: The occurrences are counted beginning with `1` and not `0`, this is different than in
```

**Review by rwst — APPROVED** [review]
The whole change is much welcome. As a new user I can understand the text, and Grammarl y only complains about "afterward(s)".

**Comment by Shreyas4991** [comment]
That's some Americanism. "Afterwards" is the correct version in English. 

See : https://dictionary.cambridge.org/thesaurus/articles/afterwards

**Review by Ruben-VandeVelde — COMMENTED** [review]
Thanks!

File: Mathlib/Tactic/NthRewrite.lean
Comment (Shreyas4991) [reply Mathlib/Tactic/NthRewrite.lean:52]: I think a bit of redundancy is better since the docstring is not a one-liner.

File: Mathlib/Tactic/NthRewrite.lean
Comment (Shreyas4991) [reply Mathlib/Tactic/NthRewrite.lean:22]: I added them and also some examples.

File: Mathlib/Tactic/NthRewrite.lean
Comment (Shreyas4991) [reply Mathlib/Tactic/NthRewrite.lean]: I added some examples which added lines to the docstring. So the redundancy helps even more.

File: Mathlib/Tactic/NthRewrite.lean
Comment (joneugster) [inline Mathlib/Tactic/NthRewrite.lean:119]: I would suggest to `import Mathlib.Tactic.ExtendDoc` and then use `@[inherit_doc nthRewriteSeq]` on the`nth_rw` macro along with the something like the following just below it:

```
extend_docs nthRwSeq before
  "`nth_rw` is like `nth_rewrite`, but also tries to close the goal by trying `rfl` afterwards.
  "
```
That way you could avoid writing that much documentation twice. The user will first see the sentense about `nth_rw` being essentially `nth_rewrite` followed by the doc for the latter.

File: Mathlib/Tactic/NthRewrite.lean
Comment (joneugster) [inline Mathlib/Tactic/NthRewrite.lean:23]:
```suggestion
of each of the `m` equalities `eqᵢ`in that order. Occurrences are counted beginning with `1`.
For example,
```
(sentence appears twice)

File: Mathlib/Tactic/NthRewrite.lean
Comment (joneugster) [inline Mathlib/Tactic/NthRewrite.lean:33]: "from the left" is not 100% correct in my opinion, maybe you could also add the following example to explain why it is the 2nd occurence and not the 1st:

```
example (a b c : Nat) : (a + b) + c = (b + a) + c := by
  nth_rewrite 2 [Nat.add_comm]
  rfl
```

File: Mathlib/Tactic/NthRewrite.lean
Comment (joneugster) [inline Mathlib/Tactic/NthRewrite.lean:57]: This part I find quite hard to read, and I do find this behaviour intuitive as is, not requiring much explanation.

Maybe this entire paragraph could be shortened to something like the following?

> `nth_rewrite` with multiple arguments recalculates the occurences after each step newly, so for `(h : a = a + b)`, `nth_rewrite 3 [h, h]` changes `a + a + a + a + a = 0` to `a + a + (a + b) + a + a = 0` in the first step and then to `a + a + ((a + b) + b) + a + a = 0` in the second.

Comment (Shreyas4991) [reply Mathlib/Tactic/NthRewrite.lean:57]: In a previous iteration I had such a short sentence. The problem is if someone is not comfortable with binder substitution and name capture issues, then it is a bit hard for them to see what is happening without a detailed example.

File: Mathlib/Tactic/NthRewrite.lean
Comment (Shreyas4991) [reply Mathlib/Tactic/NthRewrite.lean:33]: Ah yes, this is true. Technically `nth_rewrite` looks at expression trees. I am not sure how convey this intuitively without referring to the expression tree idea.

File: Mathlib/Tactic/NthRewrite.lean
Comment (Shreyas4991) [reply Mathlib/Tactic/NthRewrite.lean:119]: That sounds nice, but it doesn't adapt the examples very well. I was hoping the documentation could be entirely self-contained.

File: Mathlib/Tactic/NthRewrite.lean
Comment (Shreyas4991) [reply Mathlib/Tactic/NthRewrite.lean:35]: Fixed it with `order of precedence` and added an example to explain the importance of order of precedence.

File: Mathlib/Tactic/NthRewrite.lean
Comment (joneugster) [inline Mathlib/Tactic/NthRewrite.lean:73]:
```suggestion
⊢ a + a + (a + b + b) + a + a = 0
```
Isn't it?

File: Mathlib/Tactic/NthRewrite.lean
Comment (joneugster) [reply Mathlib/Tactic/NthRewrite.lean]: Fair enough. I also think the new iteration with the goal states is good.

File: Mathlib/Tactic/NthRewrite.lean
Comment (Shreyas4991) [reply Mathlib/Tactic/NthRewrite.lean:73]: No this is the proof state after the first rewrite.

File: Mathlib/Tactic/NthRewrite.lean
Comment (joneugster) [reply Mathlib/Tactic/NthRewrite.lean:148]: I have to admit, I didn't even read the second docstring. If it is merely a `nth_rewrite --> nth_rw` renaming, I think this suggestion would make a lot of sense and wouldn't be confusing for the reader. And it would simplify maintainability (e.g. I think the last example of the second docstring is already out-of-sync again)

However, if there is more change to it might not be ideal.

Anyways, this is a completely _optional_ suggestion!

File: Mathlib/Tactic/NthRewrite.lean
Comment (joneugster) [inline Mathlib/Tactic/NthRewrite.lean:80]: I think the remaining question from Zulip is whether you want to remove this. I personally think it's perferctly fine to drop it.

File: Mathlib/Tactic/NthRewrite.lean
Comment (joneugster) [reply Mathlib/Tactic/NthRewrite.lean]: I think I rather agree with Ruben here, but I guess that doesn't really matter either.

File: Mathlib/Tactic/NthRewrite.lean
Comment (j-loreaux) [reply Mathlib/Tactic/NthRewrite.lean:80]: I think we should drop the last paragraph. No need to remind users of Lean 3 behavior a year past the completion of the port.

**Comment by j-loreaux** [comment]
LGTM. Another maintainer can review the above unresolved conversations to see if they feel anything should be done. Make sure to wait for CI (probably doesn't matter since it's only docstrings, but still).

maintainer merge

**Comment by Shreyas4991** [comment]
what is a `maintainer merge` and given that the bot says it is already on some merge queue, I am not sure how I wait for CI.

File: Mathlib/Tactic/NthRewrite.lean
Comment (semorrison) [reply Mathlib/Tactic/NthRewrite.lean:136]: Even better, just have a single doc-string, copied via `inherit_doc` that explains both together. e.g.

`nth_rw` and `nth_rewrite` are analogues of `rw` and `rewrite` that allowing restrict which occurrence of a patttern to replace. `nth_rw` tries to close the goal after rewriting with a reducible transparency `rfl`, while `nth_rewrite` does not (just as for `rw` and `rewrite`).

.... now continue with examples just in terms of `nth_rw`.

Comment (Shreyas4991) [reply Mathlib/Tactic/NthRewrite.lean:136]: The difference is the doc tooltip only shows you the docstring for the tactic your mouse pointer is hovering on. So if I am hovering on nth_rw, I would like to see a full explanation rather than being redirected. If inherit doc could accept parameters and format strings, then the docstrings would be correct, especially the examples. My rule of thumb for documentation examples is that they should be copy pastable with minimum modification.

File: Mathlib/Tactic/NthRewrite.lean
Comment (Vierkantor) [inline Mathlib/Tactic/NthRewrite.lean:89]: I think users will be much more familiar with `rw` than `nth_rewrite`, so I would rephrase in terms of `rw`.
```suggestion
`nth_rw` is a variant of `rw` that only changes the `n`ᵗʰ _occurrence_ of the expression to be
rewritten. Like `rw`, unlike `nth_rewrite` it will try to close the goal by trying `rfl` afterwards.
```

File: Mathlib/Tactic/NthRewrite.lean
Comment (Vierkantor) [reply Mathlib/Tactic/NthRewrite.lean:136]: I think now that we have the customized docstrings, the question is whether they will be diverging, so that one is updated and the other isn't. That's potentially more harm than copypastability breaking. My estimate is that we'll be okay with two similar docstrings, so I'm inclined to keep these as-is.

**Review by Vierkantor — APPROVED** [review]
@semorrison I'm okay either way, applying your suggestion to copy the docs or not. Do you want to make the final decision?

File: Mathlib/Tactic/NthRewrite.lean
Comment (Shreyas4991) [reply Mathlib/Tactic/NthRewrite.lean:89]: This is a very good idea. I'll add it in right away.

**Comment by j-loreaux** [comment]
@Shreyas4991 a `maintainer merge` is when a [reviewer](https://leanprover-community.github.io/teams/reviewers.html) places a PR on a queue (really just a thread in a private channel on Zulip) for a maintainer to look at and potentially merge. Sometimes maintainers themselves use this queue so that they can get approval from another maintainer, as I have done here. You can still wait for CI as normal, but unlike delegation, a `maintainer merge` does not give anyone the authority to merge (other than maintainers themselves, but they already had such authority).

File: Mathlib/Tactic/NthRewrite.lean
Comment (Shreyas4991) [reply Mathlib/Tactic/NthRewrite.lean:137]: Kim  has reacted with a (y) to @Vierkantor 's suggestion to keep these strings as is. Shall I assume that this can be marked resolved?

File: Mathlib/Tactic/NthRewrite.lean
Comment (joneugster) [inline Mathlib/Tactic/NthRewrite.lean:42]:
```suggestion
Here, although the occurrence parameter is `2`, `(a + b)` is rewritten to `(b + a)`. This happens
```

File: Mathlib/Tactic/NthRewrite.lean
Comment (joneugster) [inline Mathlib/Tactic/NthRewrite.lean:90]:
```suggestion
rewritten. Like `rw`, unlike `nth_rewrite`, it will try to close the goal by trying `rfl` afterwards.
```
maybe?

File: Mathlib/Tactic/NthRewrite.lean
Comment (joneugster) [inline Mathlib/Tactic/NthRewrite.lean:109]:
```suggestion
Here, although the occurrence parameter is `2`, `(a + b)` is rewritten to `(b + a)`. This happens
```

**Review by joneugster — APPROVED** [review]
I resolved the open suggestion, since I don't consider them necessary and other reviewers didn't either.

I read through the text again and it's really clear. Left some comma suggestions, but I think it should be merged with or without them

**Review by Ruben-VandeVelde — APPROVED** [review]
maintainer merge

**Comment by j-loreaux** [comment]
bors merge

**Pull request closed without merging**
